cloudstorage/cloudstorage.go

```diff
@@ -0,0 +1,82 @@
+package cloudstorage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"fx.prodigy9.co/config"
+	"github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7/pkg/credentials"
+)
+
+var (
+	MinioAccessKeyConfig = config.Str("MINIO_ACCESS_KEY")
+	MinioSecretKeyConfig = config.Str("MINIO_SECRET_KEY")
+	MinioEndpointConfig  = config.Str("MINIO_ENDPOINT")
+
+	client *minio.Client
+)
+
+func connect(cfg *config.Source) (*minio.Client, error) {
+	if client != nil {
+		return client, nil
+	}
+
+	minioAccessKey := config.Get(cfg, MinioAccessKeyConfig)
+	minioSecretKey := config.Get(cfg, MinioSecretKeyConfig)
+	minioEndpoint := config.Get(cfg, MinioEndpointConfig)
+
+	var err error
+	client, err = minio.New(minioEndpoint, &minio.Options{
+		Creds:  credentials.NewStaticV4(minioAccessKey, minioSecretKey, ""),
+		Secure: true,
+	})
+	if err != nil {
+		client = nil
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func GeneratePresignedGetURL(bucket, objectKey string, expiry time.Duration) (string, error) {
+	client, err := connect(config.Configure())
```

Comment:
1. Sub packages should not call `config.Configure()` directly, this should be passed-in by package consumer so they can control where the configuration comes from. There maybe more possible `config.Source` implementation in the future (i.e. read data from etcd/vault dynamically)

2. `bucket` and `client` rarely changes throughout the application's lifetime. It might be worth it to implement a `*Client` wrapper or do the same thing we did with the `./data` package and put it in contextes.

```go
func PresignedGetURL(ctx context.Context, key string, age time.Duration) ...
```

4. We can drop the `Generate` part as we can think of the URLs as a "smart" getters (so we can cheat a bit and use nouns as method names, it's shorter)

The callsite should looks like this:

```go
avatarKey := "/avatars/"+user.ID+".jpg"
avatarUrl, err := objstore.PresignedGet(ctx, avatarKey, 1 * time.Hour)
```

Which already reads well in English.

5. `expiry` or expires date, is a point in time, not a duration. I think you want `age time.Duration` instead.

Reply:
fixed krub. however, would you mind checking again to see if it aligns with our recent discussion?